utils/gen-snapshot-tests.js

```diff
@@ -44,59 +177,101 @@ function generateTestFile(root, examples, componentFileMap) {
+                },
+            });
+
+            const {code} = generate(newAst);
+
+            // add snippet the list of tests that are going to be injected into
+            // the generated-snapshot-test.js file
+            tests.push(`
+
+                it("example ${exampleIndex + 1}", () => {
+                    ${code}
+                    const tree = renderer.create(example).toJSON();
+                    expect(tree).toMatchSnapshot();
+                });
+            `);
+
+            // extract import declarations from AST
+            return extractDeclarationsFromExample(exampleBody);
+        })
+        .reduce(flattenArray, [])
+        .reduce(reduceUniqueDeclarations, {});
+
+    // converts unique declarations object to a string containing all the
+    // imports
+    const transformedDeclarations = transformDeclarations(uniqueDeclarations);
+
+    // then add import declarations from other modules
+    lines.push(transformedDeclarations);
+
     if (componentFileMap) {
         for (const [componentName, filename] of Object.entries(
             componentFileMap,
         )) {
             const relFilename = path.relative(root, filename);
-            lines.push(`import ${componentName} from "./${relFilename}";`);
+
+            // include private components
+            if (!transformedDeclarations.includes(componentName)) {
+                lines.push(`import ${componentName} from "./${relFilename}";`);
+            }
         }
     }
 
     lines.push("");
     lines.push(`describe("${modName}", () => {`);
-
-    examples.forEach((example, exampleIndex) => {
-        lines.push(`    it("example ${exampleIndex + 1}", () => {`);
-
-        const ast = babylon.parse(example, {
-            plugins: ["jsx", "flow"],
-        });
-
-        const lastStatement = ast.program.body[ast.program.body.length - 1];
-
-        if (lastStatement.type !== "ExpressionStatement") {
-            throw new Error("last line should be an expression");
-        }
-
-        if (lastStatement.expression.type !== "JSXElement") {
-            throw new Error("last line should be a JSX Element");
-        }
-
-        lines.push(
-            ...example
-                .split("\n")
-                .map((line, index) =>
-                    index + 1 === lastStatement.loc.start.line
-                        ? `        const example = ${line}`
-                        : `        ${line}`,
-                ),
-        );
-
-        lines.push("        const tree = renderer.create(example).toJSON();");
-        lines.push("        expect(tree).toMatchSnapshot();");
-
-        lines.push(`    });`);
-    });
-
+    lines.push(...tests);
     lines.push("});\n");
 
     const data = lines.join("\n");
 
     const outPath = path.join(root, "generated-snapshot.test.js");
     fs.writeFileSync(outPath, data, "utf8");
+
+    const [s, ns] = process.hrtime(hrstart);
+    const ms = ns / 1000000;
+    const kind = ms > 200 ? "yellow" : "green";
```

Comment:
I put this time (200ms) here without any data to back that up, but I'm wondering if there is some data to know the "average" execution speed time for a particular script.

Reply:
Hehe, I guess you could look at the results and make a best guess? If you think this is good then that's fine!